modules/fedimint-lnv2-client/src/send_sm.rs

```diff
@@ -172,36 +170,17 @@ impl SendStateMachine {
                         *invoice.consensus_hash::<sha256::Hash>().as_ref(),
                     )),
                 )
-                .await
-            {
-                Ok(send_result) => {
-                    if contract.verify_gateway_response(&send_result) {
-                        return send_result;
-                    }
+                .await?;
 
-                    error!(
-                        ?send_result,
-                        ?contract,
-                        ?invoice,
-                        ?federation_id,
-                        ?gateway_api,
-                        "Invalid gateway response"
-                    );
-                }
-                Err(error) => {
-                    error!(
-                        ?error,
-                        ?contract,
-                        ?invoice,
-                        ?federation_id,
-                        ?gateway_api,
-                        "Error while trying to send payment via gateway"
-                    );
-                }
-            }
+            ensure!(
+                contract.verify_gateway_response(&payment_result),
+                "Invalid gateway response: {payment_result:?}"
+            );
```

Comment:
Please preserve some of the logging, e.g. through an if+error+bail

Reply:
I assumed the logging is / should be handled by util:retry?

Reply:
Good point, what I'm concerned about is that all the extra structured data (contract, invoice, etc.) won't be available anymore in logs, just some randomly formatted error string. But you are right that we should leave the logging to `retry` and there actually is an elegant way to inject all that data into the logs created by the `retry` closure: [spans](https://docs.rs/tracing/latest/tracing/span/index.html). 

Here we could do:

```rust
        util::retry("gateway-send-payment", api_networking_backoff(), || async {
            …
        })
        .instrument(info_span!(
            ?contract,
            ?invoice,
            ?federation_id,
            ?gateway_api,
        ))
        .await
        .expect("Number of retries has no limit")
```

Or alternatively add an instrument attribute to the whole function:

```rust
    #[instrument(
        skip_all,
        fields(
            ?contract,
            ?invoice,
            ?federation_id,
            ?gateway_api,
        ),
    )]
    async fn gateway_send_payment(
        gateway_api: SafeUrl,
        federation_id: FederationId,
        contract: OutgoingContract,
        invoice: LightningInvoice,
        refund_keypair: Keypair,
        context: LightningClientContext,
    ) -> Result<[u8; 32], Signature> {
```

<details><summary>More details about span log level interaction I was wondering about myself</summary>
<p>

Let me explain how span log levels interact with log statement levels in Rust's tracing framework.

The key thing to understand is that spans and events (log statements) have independent filtering, but spans can affect the visibility of events within them. Here's how it works:

```rust
use tracing::{info, debug, info_span, Level};

fn process_data() {
    // Span with INFO level
    let span = info_span!("process_data").entered();
    
    // These will be visible since their level >= span's level
    info!("starting processing");  // visible
    debug!("detailed step");      // visible
    
    // In a debug span
    let debug_span = debug_span!("detailed_processing").entered();
    info!("still visible");       // visible
    debug!("debug info");         // visible only if debug enabled
}
```

Key the interaction:

1. A span's level acts as a gate for whether the span itself is recorded/collected
2. Log statements (events) inside a span are independently filtered based on their own levels
3. If a span is disabled due to its level being too low for the current configuration, events within it may still be recorded
4. However, those events won't have the context (fields, metadata) from the filtered-out span

## Practical Example

Here's how different configurations affect visibility:

```rust
use tracing::{info, debug, info_span, Level};
use tracing_subscriber::filter::LevelFilter;

// Configure subscriber with INFO level
tracing_subscriber::fmt()
    .with_max_level(LevelFilter::INFO)
    .init();

fn example() {
    // This span will be recorded (INFO >= INFO)
    let span = info_span!("operation").entered();
    info!("visible message");     // visible
    debug!("invisible message");  // filtered out
    
    // This span won't be recorded (DEBUG < INFO)
    let detail_span = debug_span!("details").entered();
    info!("still visible");      // visible, but without debug_span context
    debug!("invisible");         // filtered out
}
```

## Best Practices

1. Set span levels to reflect their operational significance
2. Use consistent level hierarchies (don't put critical events in debug spans)
3. Remember that span levels affect metadata collection, not just visibility
4. Consider using different filters for spans vs events if needed through the `EnvFilter` or custom filters

This layered filtering system allows for flexible and granular control over your application's observability while maintaining performance.

</p>
</details>